components/raftstore-v2/src/operation/command/write/mod.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::raft_cmdpb::RaftCmdRequest;
+use raftstore::{
+    store::{
+        cmd_resp,
+        fsm::{apply, Proposal},
+        msg::ErrorCallback,
+        WriteCallback,
+    },
+    Result,
+};
+use tikv_util::Either;
+
+use crate::{batch::StoreContext, raft::Peer, router::CmdResChannel};
+
+mod raw_write;
+
+pub use raw_write::{RawWriteDecoder, RawWriteEncoder};
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    #[inline]
+    pub fn on_write_command<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: CmdResChannel,
+    ) {
+        if !self.serving() {
+            apply::notify_req_region_removed(self.region_id(), ch);
+            return;
+        }
+        if let Some(encoder) = self.raw_write_encoder_mut() {
+            match encoder.amend(req) {
+                Ok(()) => {
+                    encoder.add_response_channel(ch);
+                    self.set_has_ready();
+                    return;
+                }
+                Err(r) => req = r,
+            }
+        }
+        if let Err(e) = self.validate_command(&req, &mut ctx.raft_metrics) {
```

Comment:
admin commands need validation as well. call it `on_command`?

Reply:
It will be called when admin commands are implemented. Because encoder can batch all the writes. They will share the same header and only needs to be checked once.